File: android/lca/src/main/java/com/dreampany/lca/vm/ExchangeViewModel.java
package com.dreampany.lca.vm;

import android.app.Application;

import androidx.annotation.NonNull;

import com.dreampany.framework.data.enums.UiState;
import com.dreampany.framework.data.model.Response;
import com.dreampany.framework.misc.AppExecutors;
import com.dreampany.framework.misc.ResponseMapper;
import com.dreampany.framework.misc.RxMapper;
import com.dreampany.framework.misc.SmartMap;
import com.dreampany.framework.misc.exceptions.ExtraException;
import com.dreampany.framework.misc.exceptions.MultiException;
import com.dreampany.framework.ui.vm.BaseViewModel;
import com.dreampany.lca.api.cmc.enums.CmcCurrency;
import com.dreampany.lca.data.model.Coin;
import com.dreampany.lca.data.model.Exchange;
import com.dreampany.lca.data.source.repository.ExchangeRepository;
import com.dreampany.lca.misc.Constants;
import com.dreampany.lca.misc.CurrencyFormatter;
import com.dreampany.lca.ui.enums.TimeType;
import com.dreampany.lca.ui.model.ExchangeItem;
import com.dreampany.lca.ui.model.UiTask;
import com.dreampany.network.manager.NetworkManager;
import com.dreampany.network.data.model.Network;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.MaybeSource;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;

import javax.inject.Inject;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Created by Hawladar Roman on 6/12/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
public class ExchangeViewModel
        extends BaseViewModel<Exchange, ExchangeItem, UiTask<Coin>>
        implements NetworkManager.Callback {

    private static final int LIMIT = Constants.Limit.COIN_EXCHANGE;
    private static final long INITIAL_DELAY_IN_SECOND = 0L;
    private static final long PERIOD_IN_SECOND = 10L;
    private static final int RETRY_COUNT = 3;

    private final NetworkManager network;
    private final ExchangeRepository repo;
    private CurrencyFormatter formatter;
    private CmcCurrency cmcCurrency;
    private TimeType timeType;

    @Inject
    ExchangeViewModel(Application application,
                      RxMapper rx,
                      AppExecutors ex,
                      ResponseMapper rm,
                      NetworkManager network,
                      ExchangeRepository repo,
                      CurrencyFormatter formatter) {
        super(application, rx, ex, rm);
        this.network = network;
        this.repo = repo;
        this.formatter = formatter;
    }

    @Override
    public void clear() {
        network.deObserve(this);
        super.clear();
    }

    @Override
    public void onNetworkResult(@NonNull List<Network> networks) {

        UiState state = UiState.OFFLINE;
        for (Network network : networks) {
            if (network.getInternet()) {
                state = UiState.ONLINE;
                Response<List<ExchangeItem>> result = getOutputs().getValue();
                if (result == null || result instanceof Response.Failure) {
                    getEx().postToUi(() -> loads(false), 250L);
                }
            }
        }
        UiState finalState = state;
        getEx().postToUiSmartly(() -> updateUiState(finalState));
    }

    public void start() {
        network.observe(this, true);
    }

    public void loads(boolean important) {
        if (!takeAction(important, getMultipleDisposable())) {
            return;
        }
        Disposable disposable = getRx()
                .backToMain(getItemsRx())
                .doOnSubscribe(subscription -> postProgress(true))
                .subscribe(result -> {
                    postResult(Response.Type.ADD,result, true);
                }, error -> {
                    postFailures(new MultiException(error, new ExtraException()));
                });

        addMultipleSubscription(disposable);
    }

    private Flowable<List<ExchangeItem>> getItemsInterval() {
        return Flowable
                .interval(INITIAL_DELAY_IN_SECOND, PERIOD_IN_SECOND, TimeUnit.SECONDS, getRx().io())
                .map(tick -> getItemsRx().blockingGet())
                .retry(RETRY_COUNT);
    }

    private Maybe<List<ExchangeItem>> getItemsRx() {
        Coin coin = getTask().getInput();
        return repo
                .getItemsRx(coin.getSymbol(), LIMIT)
                .flatMap((Function<List<Exchange>, MaybeSource<List<ExchangeItem>>>) this::getItemsRx);
    }

    private Maybe<List<ExchangeItem>> getItemsRx(List<Exchange> items) {
        return Flowable.fromIterable(items)
                .map(this::getItem)
                .toList()
                .toMaybe();
    }

    private ExchangeItem getItem(Exchange exchange) {
        SmartMap<String, ExchangeItem> map = getUiMap();
        ExchangeItem item = map.get(exchange.getId());
        if (item == null) {
            item = ExchangeItem.Companion.getItem(exchange);
            map.put(exchange.getId(), item);
        }
        item.setItem(exchange);
        return item;
    }
}

File: android/crypto/src/main/java/com/dreampany/crypto/ui/home/vm/ExchangeViewModel.kt
package com.dreampany.crypto.ui.home.vm

import android.app.Application
import com.dreampany.crypto.data.enums.Action
import com.dreampany.crypto.data.enums.State
import com.dreampany.crypto.data.enums.Subtype
import com.dreampany.crypto.data.enums.Type
import com.dreampany.crypto.data.model.Exchange
import com.dreampany.crypto.data.source.repo.ExchangeRepo
import com.dreampany.crypto.misc.constants.Constants
import com.dreampany.crypto.misc.func.CurrencyFormatter
import com.dreampany.crypto.ui.home.model.ExchangeItem
import com.dreampany.framework.misc.func.ResponseMapper
import com.dreampany.framework.misc.func.SmartError
import com.dreampany.framework.ui.model.UiTask
import com.dreampany.framework.ui.vm.BaseViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by roman on 3/21/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
class ExchangeViewModel
@Inject constructor(
    application: Application,
    rm: ResponseMapper,
    private val formatter: CurrencyFormatter,
    private val repo: ExchangeRepo
) : BaseViewModel<Type, Subtype, State, Action, Exchange, ExchangeItem, UiTask<Type, Subtype, State, Action, Exchange>>(
    application,
    rm
) {

    fun loadExchanges(fromSymbol: String, toSymbol: String, extraParams: String) {
        uiScope.launch {
            postProgressMultiple(true)
            var result: List<Exchange>? = null
            var errors: SmartError? = null
            try {
                result = repo.getExchanges(
                    fromSymbol,
                    toSymbol,
                    extraParams,
                    Constants.Limits.EXCHANGES
                )
            } catch (error: SmartError) {
                Timber.e(error)
                errors = error
            }
            if (errors != null) {
                postError(errors)
            } else {
                postResult(result?.toItems())
            }
        }
    }

    private suspend fun List<Exchange>.toItems(): List<ExchangeItem> {
        val input = this
        return withContext(Dispatchers.IO) {
            input.map { input ->
                ExchangeItem.getItem(input, formatter)
            }
        }
    }

    private fun postProgressMultiple(progress: Boolean) {
        postProgressMultiple(
            Type.EXCHANGE,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            progress = progress
        )
    }


    private fun postError(error: SmartError) {
        postMultiple(
            Type.EXCHANGE,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            error = error,
            showProgress = true
        )
    }

    private fun postResult(result: List<ExchangeItem>?) {
        postMultiple(
            Type.EXCHANGE,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            result = result,
            showProgress = true
        )
    }
}